auth-sessions/app.js

```diff
@@ -0,0 +1,232 @@
+ * limitations under the License.
+ */
+
+var express = require('express');
+var cookieParser = require('cookie-parser')
+var http = require('http');
+var https = require('https');
+var app = express();
+var admin = require('firebase-admin');
+var bodyParser = require('body-parser');
+
+/**
+ * Renders the profile page and serves it in the response.
+ * @param {string} endpoint The get profile endpoint.
+ * @param {!Object} req The expressjs request.
+ * @param {!Object} res The expressjs response.
+ * @param {!firebase.auth.DecodedIdToken} decodedClaims The decoded claims from verified
+ *     session cookies.
+ * @return {!Promise} A promise that resolves on success.
+ */
+function serveContentForUser(endpoint, req, res, decodedClaims) {
+  // Lookup the user information corresponding to cookie and return the profile data for the user.
+  return admin.auth().getUser(decodedClaims.sub).then(function(userRecord) {
+    var html = '<!DOCTYPE html>' +
+      '<html>' +
+      '<meta charset="UTF-8">' +
+      '<link href="style.css" rel="stylesheet" type="text/css" media="screen" />' +
+      '<meta name="viewport" content="width=device-width, initial-scale=1">' +
+      '<title>Sample Profile Page</title>' +
+      '<body>' +
+      '<div id="container">' +
+      '  <h3>Welcome to My Awesome App, '+( userRecord.displayName || 'N/A') +'</h3>' +
+      '  <div id="loaded">' +
+      '    <div id="main">' +
+      '      <div id="user-signed-in">' +
+      // Show user profile information.
+      '        <div id="user-info">' +
+      '          <div id="photo-container">' +
+      (userRecord.photoURL ? '      <img id="photo" src=' +userRecord.photoURL+ '>' : '') +
+      '          </div>' +
+      '          <div id="name">' + userRecord.displayName + '</div>' +
+      '          <div id="email">'+
+      userRecord.email + ' (' + (userRecord.emailVerified ? 'verified' : 'unverified') + ')</div>' +
+      '          <div class="clearfix"></div>' +
+      '        </div>' +
+      '        <p>' +
+      // Append button for sign out.
+      '          <button id="sign-out" onClick="window.location.assign(\'/logout\')">Sign Out</button>' +
+      // Append button for deletion.
+      '          <button id="delete-account" onClick="window.location.assign(\'/delete\')">' +
+      'Delete account</button>' +
+      '        </p>' +
+      '      </div>' +
+      '    </div>' +
+      '  </div>' +
+      '</div>' +
+      '</body>' +
+      '</html>';
+    res.set('Content-Type', 'text/html');
+    res.end(html);
+  });
+}
+
+/**
+ * Attaches a CSRF token to the request.
+ * @param {string} url The URL to check.
+ * @param {string} cookie The CSRF token name.
+ * @param {string} value The CSRF token value to save.
+ * @return {function} The middleware function to run.
+ */
+function attachCsrfToken(url, cookie, value) {
+  return function(req, res, next) {
+    if (req.url == url) {
+      res.cookie(cookie, value);
+    }
+    next();
```

Comment:
I can't find the definition of this function. Is this an Express built-in?

Reply:
https://expressjs.com/en/guide/writing-middleware.html